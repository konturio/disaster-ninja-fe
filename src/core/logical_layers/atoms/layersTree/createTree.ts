import type { CategorySettings, GroupSettings, Tree } from './types';

export function createTree(
  layers: { id: string; category?: string; group?: string; order?: number }[],
  {
    categoriesSettings,
    groupsSettings,
  }: {
    categoriesSettings: { [x: string]: CategorySettings };
    groupsSettings: { [x: string]: GroupSettings };
  },
) {
  const categories = new Map();
  const groups = new Map();
  // Categories may have groups with same names
  const getUniqueGroupKey = (layer: { category?: string; group?: string }) =>
    layer.category ? `${layer.category}__${layer.group}` : layer.group;

  const tree: Tree = {
    children: [],
  };

  layers.forEach((layer) => {
    if (!layer.group) {
      // It's a root layer, because layer without group can't be children of category too
      tree.children.push(layer);
      return;
    }

    /* Create category if not exist */
    if (layer.category && !categories.has(layer.category)) {
      const categorySettings = categoriesSettings[layer.category] ?? {};
      const category = {
        ...categorySettings,
        isCategory: true as const,
        id: layer.category,
        children: [],
      };
      categories.set(layer.category, category);
      tree.children.push(category);
    }

    /* Create group if not exist */
    const groupKey = getUniqueGroupKey(layer);
    if (!groups.has(groupKey)) {
      const groupSettings = groupsSettings[layer.group] ?? {};
      const group = {
        ...groupSettings,
        isGroup: true as const,
        id: layer.group,
        children: [],
      };
      groups.set(groupKey, group);
      if (layer.category) {
        const category = categories.get(layer.category);
        category.children.push(group);
      } else {
        tree.children.push(group);
      }
    }

    /* Add layer to group */
    const group = groups.get(groupKey);
    group.children.push(layer);
  });

  return tree;
}
