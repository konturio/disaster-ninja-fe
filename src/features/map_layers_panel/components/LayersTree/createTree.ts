import { Tree } from './types';

export function createTree(layers): Tree {
  return layers;
}
