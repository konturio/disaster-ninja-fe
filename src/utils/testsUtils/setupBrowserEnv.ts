global['window'] = { konturAppConfig: {} } as any;

// need this to convert ts file to esmodule which is needed by esbuild
export {};
