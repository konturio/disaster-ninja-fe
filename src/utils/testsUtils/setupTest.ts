import anyTest, { TestInterface } from 'ava';

/* Allow setup typed context */
// https://github.com/avajs/ava/blob/main/docs/recipes/typescript.md
export function setupTestContext<T extends Record<string, unknown>>(
  contextOrCallback: (() => T) | T,
) {
  let context: T;
  const test = anyTest as TestInterface<typeof context>;
  test.beforeEach((t) => {
    if (typeof contextOrCallback === 'function') {
      context = contextOrCallback();
    } else {
      context = contextOrCallback;
    }
    t.context = context;
  });
  return test;
}

export const createLocalStorageMock = () => ({
  getItem: (key: string) => null,
  setItem: (key: string, value: string) => null,
  removeItem: (key: string) => null,
  clear: () => null,
  key: (idx: number) => null,
  length: 0,
});
