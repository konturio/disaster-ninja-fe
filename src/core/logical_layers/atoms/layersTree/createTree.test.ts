import { setupTestContext } from '../../../../utils/testsUtils/setupTest';
import { createTree } from './createTree';

const fakeAtom = {} as any;
const test = setupTestContext(() => {
  return {
    settings: {
      categoriesSettings: {},
      groupsSettings: {},
    },
  };
});

test('Empty registry', (t) => {
  t.deepEqual(createTree([], t.context.settings), { children: [] });
});

test('Root layer', (t) => {
  t.deepEqual(
    createTree([{ id: 'rootLayer', atom: fakeAtom }], t.context.settings),
    {
      children: [{ id: 'rootLayer', atom: fakeAtom }],
    },
  );
});

test('Layer in group', (t) => {
  t.like(
    createTree(
      [
        { id: 'layer_1', group: 'foo', atom: fakeAtom },
        { id: 'layer_2', group: 'bar', atom: fakeAtom },
        { id: 'layer_3', group: 'foo', atom: fakeAtom },
      ],
      t.context.settings,
    ),
    {
      children: [
        {
          id: 'foo',
          isGroup: true,
          children: [
            { id: 'layer_1', group: 'foo', atom: fakeAtom },
            { id: 'layer_3', group: 'foo', atom: fakeAtom },
          ],
        },
        {
          id: 'bar',
          isGroup: true,
          children: [{ id: 'layer_2', group: 'bar', atom: fakeAtom }],
        },
      ],
    },
  );
});

test('Layer in category and group', (t) => {
  const tree = createTree(
    [
      {
        id: 'layer_1',
        group: 'foo_group',
        category: 'foo_category',
        atom: fakeAtom,
      },
    ],
    t.context.settings,
  );
  t.true(tree.children[0].id === 'foo_category');
  // @ts-expect-error it's ok to fail this test if property undefined
  t.true(tree.children[0].children[0].id === 'foo_group');
  // @ts-expect-error it's ok to fail this test if property undefined
  t.true(tree.children[0].children[0].children[0].id === 'layer_1');
});

test('Integration - create correct tree', (t) => {
  t.snapshot(
    createTree(
      [
        { id: 'rootLayer', atom: fakeAtom },
        {
          id: 'layer_1',
          group: 'foo_group',
          category: 'foo_category',
          atom: fakeAtom,
        },
        {
          id: 'layer_2',
          group: 'foo_group',
          category: 'foo_category',
          atom: fakeAtom,
        },
        {
          id: 'layer_3',
          group: 'foo_group',
          category: 'bar_category',
          atom: fakeAtom,
        },
        {
          id: 'layer_4',
          group: 'bar_group',
          category: 'bar_category',
          atom: fakeAtom,
        },
        { id: 'layer_5', group: 'bar_group', atom: fakeAtom },
      ],
      t.context.settings,
    ),
  );
});

test('Tree apply group and category settings', (t) => {
  t.like(
    createTree(
      [
        {
          id: 'layer_1',
          group: 'foo_group',
          category: 'foo_category',
          atom: fakeAtom,
        },
      ],
      {
        categoriesSettings: {
          foo_category: {
            name: 'Foo category',
            order: 3,
            openByDefault: false,
            mutuallyExclusive: false,
          },
        },
        groupsSettings: {
          foo_group: {
            name: 'Foo group',
            order: 5,
            openByDefault: true,
            mutuallyExclusive: true,
          },
        },
      },
    ),
    {
      children: [
        {
          id: 'foo_category',
          name: 'Foo category',
          isCategory: true,
          order: 3,
          openByDefault: false,
          mutuallyExclusive: false,
          children: [
            {
              id: 'foo_group',
              name: 'Foo group',
              isGroup: true,
              order: 5,
              openByDefault: true,
              mutuallyExclusive: true,
              children: [
                {
                  id: 'layer_1',
                  group: 'foo_group',
                  category: 'foo_category',
                  atom: fakeAtom,
                },
              ],
            },
          ],
        },
      ],
    },
  );
});
