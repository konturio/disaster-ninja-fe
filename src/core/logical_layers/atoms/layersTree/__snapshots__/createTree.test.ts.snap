// Vitest Snapshot v1

exports[`Integration - create correct tree 1`] = `
{
  "children": [
    {
      "atom": {},
      "id": "rootLayer",
    },
    {
      "children": [
        {
          "children": [
            {
              "atom": {},
              "category": "foo_category",
              "group": "foo_group",
              "id": "layer_1",
            },
            {
              "atom": {},
              "category": "foo_category",
              "group": "foo_group",
              "id": "layer_2",
            },
          ],
          "id": "foo_group",
          "isGroup": true,
        },
      ],
      "id": "foo_category",
      "isCategory": true,
    },
    {
      "children": [
        {
          "children": [
            {
              "atom": {},
              "category": "bar_category",
              "group": "foo_group",
              "id": "layer_3",
            },
          ],
          "id": "foo_group",
          "isGroup": true,
        },
        {
          "children": [
            {
              "atom": {},
              "category": "bar_category",
              "group": "bar_group",
              "id": "layer_4",
            },
          ],
          "id": "bar_group",
          "isGroup": true,
        },
      ],
      "id": "bar_category",
      "isCategory": true,
    },
    {
      "children": [
        {
          "atom": {},
          "group": "bar_group",
          "id": "layer_5",
        },
      ],
      "id": "bar_group",
      "isGroup": true,
    },
  ],
}
`;
